cgsmiles/sample.py

```diff
@@ -0,0 +1,139 @@
+import numpy as np
+import networkx as nx
+from .read_fragments import read_fragments
+from .graph_utils import (merge_graphs,
+                          sort_nodes_by_attr,
+                          set_atom_names_atomistic)
+from .pysmiles_utils import rebuild_h_atoms
+
+def _find_complementary_bonding_descriptor(bonding_descriptor):
+    if bonding_descriptor[0] == '<':
+        compl = '>' + bonding_descriptor[1:]
+    elif bonding_descriptor[0] == '>':
+        compl = '<' + bonding_descriptor[1:]
+    else:
+        compl = bonding_descriptor
+    return compl
+
+class MoleculeSampler:
+    """
+    Given a fragment string in CGSmiles format
+    and probabilities for residues to occur,
+    return a random molecule with target
+    molecular weight.
+    """
+    def __init__(self,
+                 pattern,
+                 target_weight,
+                 bonding_probabilities,
+                 start=None,
+                 all_atom=True):
+
+        """
+        Parameters
+        ----------
+        pattern: str
+            the cgsmiles string to resolve
+        target_length: int
+            the target molecular weight in unit
+            of number of monomers
+        fragment_probabilities: dict[str, float]
+            dictionary connecting probability
+            to fragment names
+        bonding_probabilities: dict[str, float]
+            probabilty that two bonding descriptors
+            will connect
+        all_atom: bool
+            if the fragments are all-atom
+            resolution
+        """
+        self.molecule = nx.Graph()
+        self.bonding_probabilities = bonding_probabilities
+        self.all_atom = all_atom
+        self.target_weight = target_weight
+        self.start = start
+        self.current_open_bonds = defaultdict(list)
+
+        fragment_strings = re.findall(r"\{[^\}]+\}", pattern)
+
+        if len(fragment_strings) > 1:
+            raise IOError("Sampling can only be done on one resolution.")
+
+        self.fragment_dict = {}
+        self.fragment_dict.update(read_fragments(fragment_strings[0],
+                                                 all_atom=all_atom))
+
+        # we need to store which bonding descriptors is present in which
+        # fragment so we can later just look them up
+        self.fragments_by_bonding = defaultdict(list)
+        for fragname, fraggraph  in self.fragment_dict.items():
+            bondings = nx.get_node_attributes(fraggraph, "bonding")
+            for node, bondings in bondings.items():
+                for bonding in bondings:
+                    self.fragments_by_bonding[bonding].append((fragname, node))
+
+    def update_open_bonds(self):
+        self.current_open_bonds = defaultdict(list)
+        open_bonds = nx.get_node_attributes(self.molecule, 'bonding')
+        for node, bonding_types in open_bonds.items():
+            for bonding_types in bonding_types:
+                self.current_open_bonds[bonding_types].append(node)
```

Comment:
Is it worth abstracting this to a helper function?

Reply:
I guess could be useful; why not